crates/rome_service/src/file_handlers/javascript.rs

```
@@ -260,3 +263,34 @@ fn format_on_type(
     let printed = rome_js_formatter::format_sub_tree(context, &root_node)?;
     Ok(printed)
 }
+
+fn rename(
+    _rome_path: &RomePath,
+    parse: AnyParse,
+    symbol_at: TextSize,
+    new_name: String,
+) -> Result<String, RomeError> {
+    use rome_js_analyze::utils::rename::RenameSymbolExtensions;
+
+    let root = parse.tree();
+    let model = semantic_model(&root);
+
+    if let Some(node) = parse
+        .syntax()
+        .descendants_tokens(Direction::Next)
+        .find(|token| token.text_range().contains(symbol_at))
+        .and_then(|token| token.parent())
+    {
+        match node.try_into() {
+            Ok(node) => {
+                let mut batch = root.begin();
+                batch.rename_with_any_can_be_renamed(&model, node, &new_name);
```

Should we return an error if the renaming fails in case `new_name` is already used?

---

I found that vscode will not emit an error when the new name is already used.

---

Yes, we should. Done.